src/lib/cpd/CpdWrapper.ts

```diff
@@ -0,0 +1,98 @@
+import {Logger} from '@salesforce/core';
+import {FileHandler} from '../util/FileHandler';
+import {ChildProcessWithoutNullStreams} from 'child_process';
+import { PmdSupport} from '../pmd/PmdSupport';
+import * as JreSetupManager from './../JreSetupManager';
+import path = require('path');
+
+const MAIN_CLASS = 'net.sourceforge.pmd.cpd.CPD';
+const HEAP_SIZE = '-Xmx1024m';
+
+interface CpdWrapperOptions {
+	path: string;
+	language: string;
+	minimumTokens: number;
+}
+
+export default class CpdWrapper extends PmdSupport {
+
+	path: string;
+	language: string;
+	minimumTokens: number;
+	logger: Logger; 
+	private initialized: boolean;
+
+	protected async init(): Promise<void> {
+
+		if (this.initialized) {
+			return;
+		}
+
+		this.logger = await Logger.child('CpdWrapper');
+		this.initialized = true;
+	}
+
+	protected async buildCommandArray(): Promise<[string, string[]]> {
+		const javaHome = await JreSetupManager.verifyJreSetup();
+		const command = path.join(javaHome, 'bin', 'java');
+
+		const classpath = await this.buildClasspath();
+
+		const fileHandler = new FileHandler();
+		const tmpPath = await fileHandler.tmpFileWithCleanup();
+		await fileHandler.writeFile(tmpPath, this.path);
+
+		const args = ['-cp', classpath.join(path.delimiter), HEAP_SIZE, MAIN_CLASS, '--filelist', tmpPath,
+			'--format', 'xml', '--minimum-tokens', String(this.minimumTokens), '--language', this.language];
+
+		this.logger.trace(`Preparing to execute CPD with command: "${command}", args: "${args}"`);
+		return [command, args];
+	}
+
+	protected monitorChildProcess(cp: ChildProcessWithoutNullStreams, res: (string) => void, rej: (string) => void): void {
```

Comment:
From a quick look, this looks identical to PmdWrapper's `monitorChildProcess`. Can this method be moved to PmdSupport so that both implementations can reuse the same code without duplication?

Reply:
The only issue is that the method needs access to the engine's logger attribute, but I could pass that into the method and define it in a higher level.

Reply:
It's being called from RunCommand, so I'm not sure I can do this.